dra-evolution/pkg/api/capacity_types.go

```diff
@@ -1,170 +1,162 @@
-	Conditions []metav1.Condition `json:"conditions"`
-
-	// ClaimUIDs contains the UIDs of the claims to which this device
-	// is allocated.
-	ClaimUIDs []types.UID
+// NamedDevices is used in ResourceModel.
+type NamedDevices struct {
+	// The list of all devices currently available.
+	//
+	// +listType=atomic
+	Devices []NamedDevice `json:"devices" protobuf:"bytes,1,name=instances"`
 }
 
-// Device is used to track individual devices in a pool.
-type Device struct {
-	// Name is a driver-specific identifier for the device.
-	// +required
-	Name string `json:"name"`
+// NamedDevice represents one individual hardware instance that can be selected based
+// on its attributes.
+type NamedDevice struct {
+	// Name is unique identifier among all devices managed by
+	// the driver on the node. It must be a DNS subdomain.
+	Name string `json:"name" protobuf:"bytes,1,name=name"`
 
-	// Attributes contain additional metadata that can be used in
-	// constraints. If an attribute name overlaps with the pool attribute,
-	// the device attribute takes precedence.
-	//
-	// +optional
-	Attributes []Attribute `json:"attributes,omitempty"`
-
-	// SharedResourcesConsumed contains the pooled resources that are
-	// consumed when this device is allocated.
+	// Attributes defines the attributes of this device.
+	// The name of each attribute must be unique.
 	//
+	// +listType=atomic
 	// +optional
-	SharedResourcesConsumed map[string]resource.Quantity `json:"sharedResourcesConsumed,omitempty"`
+	Attributes []DeviceAttribute `json:"attributes,omitempty" protobuf:"bytes,3,opt,name=attributes"`
 
-	// ClaimResourcesProvided allows the definition of per-device resources
-	// that can be allocated in a manner similar to standard Kubernetes
-	// resources.
-	//
-	// +optional
-	ClaimResourcesProvided []ResourceCapacity `json:"claimResourcesProvided,omitempty"`
+	// TODO for 1.31: define how to support partitionable devices
 }
 
-type ResourceCapacity struct {
-	// Name is the resource name/type.
-	// +required
-	Name string `json:"name"`
-
-	// Capacity is the total capacity of the named resource.
-	// +required
-	Capacity resource.Quantity `json:"capacity"`
-
-	// BlockSize is the increments in which capacity is consumed. For
-	// example, if you can only allocate memory in 4k pages, then the
-	// block size should be "4Ki". Default is 1.
+// DeviceAttribute is a combination of an attribute name and its value.
+type DeviceAttribute struct {
+	// Name is a unique identifier across all possible attributes of devices.
 	//
-	// If the resource is consumable in a fractional way, then the
-	// default of 1 should not be used; instead this should be a fractional
-	// amount corresponding the increment size. We may also need a minimum
-	// value, if the minimum is larger than the block size (as is the case
-	// for standard Kubernetes CPU resources).
+	// If this is a DNS subdomain (no dot), then the driver name gets added
```

Comment:
Also: if we are going to use these from CEL as `foo == "xyz" && bar > 6` rather than `attributes.strings["foo"] == "xyz" && attributes.ints["bar"] > 6`, then the names probably need to be more like a C-identifier, which excludes dashes but allows underscores.

How the CEL looks has ripple.

Reply:
We will compare the attributes across devices, so we would want to standardize those. The only reason for qualification would be to avoid conflicts with the standard set. We could instead make the standard set actual field values, then have a an array of attributes that you refer to via the attributes map-like reference. 

Standard set would include driver, vendor, model and similar, plus pcieRoot etc.

Reply:
> Also, "DNS subdomain" is "foo.bar.bat". "DNS label" is "foo" or "bar" but never "foo.bar".

Sorry, my bad. I meant "DNS label". I agree that if we want this to be usable as a CEL field name, we need to validate a bit differently. We probably should not allow leading digits (allowed in DNS label) and allow underscores (not allowed in DNS label).

I had not planned to support a standard set with short names. Instead, I envisioned "driver.k8s.io" as the name for a standardized attribute (whatever it means) and there would be no way to reference them as fields in CEL.

We can allow standardization of short names. For that, we need to change the definition of `Name` here so that `foo` (no dot) does not automatically get scoped with the driver name. The implication is that all attributes defined by the vendor would have to have a suitable suffix (a DNS subdomain?), to avoid name clashes.

@klueska: you said that you don't want to be forced to publish the attributes with long names. That's where this automatism in the current proposal comes from. Would be okay if we drop it?

Reply:
So I, as the driver vendor, would need to name all of my driver-specific attributes with my driver name as follows:
```
  - name: vendor.gpu.nvidia.com
    string: nvidia
  - name: product-name.gpu.nvidia.com
    string: Mock NVIDIA A100-SXM4-40GB
  - name: brand.gpu.nvidia.com
    string: Nvidia
  - name: architecture.gpu.nvidia.com
    string: Ampere
  - name: cuda-compute-capability.gpu.nvidia.com
    version: 8.0.0
  - name: driver-version.gpu.nvidia.com
    version: 550.54.15
  - name: cuda-driver-version.gpu.nvidia.com
    version: 12.4.0
  - name: memory.gpu.nvidia.com
    quantity: 40Gi
  - name: multiprocessors.gpu.nvidia.com
    quantity: "98"
  - name: copy-engines.gpu.nvidia.com
    quantity: "7"
  - name: decoders.gpu.nvidia.com
    quantity: "5"
  - name: encoders.gpu.nvidia.com
    quantity: "0"
  - name: jpeg-engines.gpu.nvidia.com
    quantity: "1"
  - name: ofa-engines.gpu.nvidia.com
    quantity: "1"
  - name: memory-slices-0.gpu.nvidia.com
    quantity: "1"
  - name: memory-slices-0.gpu.nvidia.com
    quantity: "1"
  - name: memory-slices-1.gpu.nvidia.com
    quantity: "1"
  - name: memory-slices-2.gpu.nvidia.com
    quantity: "1"
  - name: memory-slices-3.gpu.nvidia.com
    quantity: "1"
  - name: memory-slices-4.gpu.nvidia.com
    quantity: "1"
  - name: memory-slices-5.gpu.nvidia.com
    quantity: "1"
  - name: memory-slices-6.gpu.nvidia.com
    quantity: "1"
```

I personally like the approach of *not* including this for driver-specific attributes, but *do* include it for standardized attributes.

Reply:
Except that this won't work when "driver/vendor/etc." are standardized attributes without a suffix. I think it's fair to make vendors do a bit more work here when the end result is simpler (no automatism) and more flexible (shorthand standardized attributes).

Regarding making standardized attributes separate fields: I'd prefer to avoid that, because it implies that schedulers must be updated to see these fields and make them usable. When they are listed here together with other attributes, they are available immediately also with older schedulers (version skew again).

Making a new standardized attribute available also as a CEL field has to be stretched out over two releases: in 1.X we add a new standardized attribute and add it to the CEL environment, but don't allow new expressions to use it because downgrading would render the new expression invalid. In 1.X+1, new expressions are allowed to use it because downgrading to 1.X supports it for stored expressions. See https://pkg.go.dev/k8s.io/apiserver/pkg/cel/environment#Type

There's one downside: I don't see a good way to validate standardized attributes and to prevent vendors from using "shorthand" names for their own attributes. If in 1.30 we don't allow any shorthand name because we haven't standardized one yet, add one shorthand in 1.31, and then someone downgrades, the ResourcePool becomes invalid.

We also cannot add validation of "known" attributes in 1.31, because in 1.30 a vendor might have (incorrectly!) used a shorthand name with a type that doesn't match the one we choose for 1.31, rending the old ResourcePool invalid.

Hmm, perhaps "immediately usable" isn't that important? Then separate fields work. But in that case I am not sure how to add a standardized field: as soon as we add it, it can be set, even if it does not work yet. I guess that's okay? Downgrade and you loose functionality...

Reply:
I don't see the advantage of short-hand standardized attributes if device-specific attributes still have to be fully-qualified in CEL expressions.

Reply:
In CEL expressions, "fully-qualified" would not be required for standardized attributes. It can be referenced with its short name and potentially as a field in the `device` variable.

Reply:
I understand the implications. I'm just not convinced that selecting / constraining on the "standard" attributes will be more common than doing so on device specific ones (meaning that you end up with a bunch of CEL expressions using the long-form syntax anyway).

Reply:
I tend to agree. But if we don't design for shorthand attributes now, it'll be hard to add them later.

Reply:
> We will compare the attributes across devices, so we would want to standardize those

I see how topology attributes might want to be standardized, but I am skeptical
that there's a need or a will to standardize beyond that.

That said, if we do ANY we have to spec it.  I don't love the idea that every attribute is denoted with sa vendor suffix.  Nobody wants to type that.

Let's brainstorm - just drop ideas:

TH1) Define one standard attribute "std" which is a map-like / struct-like type.  CEL would say something like `"dev.std.pcie_root == ..."`.  In fact, if these are JUST topology, do they even need to be matchable in CEL?  We could say `matchAttributes: std.pcieRoot` or `matchAttributes: topology.pcieRoot`

TH2) Allow attributes to have dots in their name and when we publish to CEL, we just treat them like sub-structs

TH3) Make standardized attributes use a suffix of "k8s.io" (or something) and those are the only ones allowed to have dots in the name, then follow TH2.

add more...

Reply:
> I don't love the idea that every attribute is denoted with sa vendor suffix. Nobody wants to type that.

If we have some CEL expression with short names, how does the code which looks up those names know which driver those names are supposed to be for? Do we want to bring back some "default" suffix that gets added when nothing is specified? I had that in some earlier revision and Kevin and I concluded that this was simply confusing an error prone.

Kevin mentioned matching attributes from different vendors in the same expression. This seems plausible, in particular when we don't standardize attribute names already in 1.31. So we still need fully-qualified names in that case because we cannot have two different defaults.

> TH1) Define one standard attribute "std" which is a map-like / struct-like type. CEL would say something like "dev.std.pcie_root == ...". In fact, if these are JUST topology, do they even need to be matchable in CEL? We could say matchAttributes: std.pcieRoot or matchAttributes: topology.pcieRoot

This simplifies "standardized" attributes, but doesn't really help with access to vendor-provided attributes.

> TH2) Allow attributes to have dots in their name and when we publish to CEL, we just treat them like sub-structs

So instead of `device.stringAttributes["driver_name.dra.example.com"]` we allow writing `device.com.example.dra.driver_name` or `device.driver_name.dra.example.com`? I very worried about field name collisions and not sure whether this really makes CEL expressions easier.

> TH3) Make standardized attributes use a suffix of "k8s.io" (or something) and those are the only ones allowed to have dots in the name, then follow TH2.

This has the problem of not dealing with conflicting vendor attributes.

Reply:
> Do we want to bring back some "default" suffix that gets added when nothing is specified?

For reference, here's what it [looked like earlier](https://github.com/kubernetes-sigs/wg-device-management/blob/45d14ba43414be9e9e2288ea160ac6eeca051f4a/dra-evolution/pkg/api/claim_types.go#L112-L164) (look for `DriverName`, which is not in the current proposal) and here's [the discussion](https://github.com/kubernetes-sigs/wg-device-management/pull/14#discussion_r1609035192) about removing it.

I don't think that looking up a qualified named as if it was a nested struct in CEL will work well. For both `device.com.example.dra.driver_name` and `device.driver_name.dra.example.com` the code which implements the lookup won't know when to stop and return `nil`.

Remember that Kubernetes uses methods like `isGeaterThan` instead of comparison operators for quantities (https://kubernetes.io/docs/reference/using-api/cel/#kubernetes-quantity-library).

So the CEL expression would be `device.com.example.dra.memory.isGreaterThan(quantity("1Gi")`. When evaluating this for a device which has no such attribute, the lookup will have to continue as if `isGreaterThan` was part of the attribute name and throw a confusing error about "attribute isGreaterThan.memory.dra.example.com not found".

Using the array syntax (`device.quantityAttributes["memory.dra.example.com"]`) doesn't have that problem, is type-safe at compile time, and returns a sane default value (`quantity("0")`) when not set for a device. I have accepted that some users may want to write `device.attributes["memory.dra.example.com"]` instead (not type-safe at compile time, nil as default), so that is part of the current proposal as an alternative. Both are also implemented in `pkg/schedule`, so I am confident that we won't run into unexpected surprises during the implementation phase.

This struct-style lookup may look nicer (although even that is probably subjective - I don't like it), but it's a major foot gun in terms of usability. I don't think we should give that to our users.

To be clear, for short names it's okay. That's a single level struct lookup.

But for those we still need to figure out how to publish them. The only approach that I see right now is the one I mentioned [earlier in this thread](https://github.com/kubernetes-sigs/wg-device-management/pull/14/files#discussion_r1615925529):
- fully-qualified names in ResourcePool when it's a vendor attribute
- short name in ResourcePool when it's defined by Kubernetes
- lookup via that name in CEL with array syntax, in addition as field when its a standard attribute

If we want to simplify the CEL expressions, we can bring back the `DriverName` field as an additional criteria and default suffix.

Reply:
This "fully-qualified names in ResourcePool when it's a vendor attribute" irks me. This makes the common case hard and the uncommon one simple.

How about we reverse it:
- short names in ResourcePool inherit the driver name as suffix
- standard attributes have `k8s.io` as suffix in ResourcePool
- lookup with suffix in array for both (current proposal) or with the automagical addition of the `DriverName` as suffix for short names (https://github.com/kubernetes-sigs/wg-device-management/pull/14#discussion_r1609035192)
- lookup of `k8s.io` attributes as `device.std.<short name>`, with CEL environment versioning as explained in https://github.com/kubernetes-sigs/wg-device-management/pull/14#discussion_r1615925529

Reply:
Slightly related: after the removal of `DriverName`, the `DeviceFilter` struct only contains a single string, the selector. If we keep that approach, then we can simplify `device: selector: "<CEL>"` to `deviceSelector: <CEL>` because we cannot add new fields to that struct anyway (would be ignored by older clients). If we need to extend, we can use the outer one-of and add `device` with a struct back as alternative to the simpler `deviceSelector`.

Reply:
> TH1) Define one standard attribute "std" which is a map-like / struct-like type. CEL would say something like "dev.std.pcie_root == ...". In fact, if these are JUST topology, do they even need to be matchable in CEL? We could say matchAttributes: std.pcieRoot or matchAttributes: topology.pcieRoot

I don't think they would just be topology. Vendor and model would be the other two I can see folks using. Also there may be *vendor defined topology* ("two migs on the same physical card").

I do like though separating out the "standard" and "vendor" attributes into two different fields, though. That means that we don't really have to qualify them, doesn't it? We know in any given device `device.vendorAttributes` only have meaning with other devices from the same vendor.

So something like `device.pcieRoot` or `device.std.pcieRoot`, and for vendor attributes something like `device.vendor.someVendorAttribute` or `device.vendor.attributes["foo"]` or `device.vendor.stringAttributes["bar"]`, etc.?

Reply:
> If we have some CEL expression with short names, how does the code which looks up those names know which driver those names are supposed to be for

I don't follow.  At the time we are considering whether a device fits this claim or not, we already have a device in mind, and so we have a total set of the attributes.  If you care about which vendor, you should make your first criterion `device.driver == "example.com"`.  If you don't care about driver, then you clearly cannot match on vendor-defined attributes.

The way I see it is that the attributes are approximately a `map[string]any` where a very small set of keys (e.g. "driver", "std") are well-defined and everything else is whatever the vendor of that device wants to publish.  We should carve out a few that we are likely to need, and the rest are vendor.  Not "device.vendor.foo", just "device.foo".  Or even "foo".

> Kevin mentioned matching attributes from different vendors in the same expression.

If two vendors agree on the same name to mean the same thing, that's fine.  But that doesn't need *us* to be involved, does it?

> This simplifies "standardized" attributes, but doesn't really help with access to vendor-provided attributes.

There are only standard attributes and vendor attributes.  You don't need to namespace vendor attributes (in this idea)

> So instead of device.stringAttributes["driver_name.dra.example.com"] we allow writing device.com.example.dra.driver_name or device.driver_name.dra.example.com? I very worried about field name collisions and not sure whether this really makes CEL expressions easier.

I don't think most fields want to be prefixed, so the majority of expressions would be `device.foo` or `device.bar` and occasionally `device.pcie.k8s.io`.  but as I write it out I like it less and less.  This is why it's a brainstorm! :)

> Vendor and model would be the other two I can see folks using.

Why do we need to standardize those?  Are we going to dictate that they exist and reject devices that don't provide them?  Do we get to define what makes a "model"?  Is it a silicon stepping?  Or a marketing name?  Or ...

> there may be vendor defined topology

Hypothetically (not sure I like it) topology attribs could be segregated from selectable attribs in the pool structure.  Or they could be `std.topo.pci` or even `topo.pcie` ?

Reply:
> Why do we need to standardize those? Are we going to dictate that they exist and reject devices that don't provide them? Do we get to define what makes a "model"? Is it a silicon stepping? Or a marketing name? Or ...

Model isn't important, but I think driver and vendor make sense, from the point of view that a vendor may define vendor-specific attributes that they want to work across multiple drivers from the same vendor. If vendor is standard, then users can rely on using it in a CEL expression to narrow to the set of devices where these are valid. Of cource they could also say `driver in list-of-drivers` so maybe that's good enough.

For match attributes, if we wanted devices of all the same vendor, model, and version, but didn't care about the vendor or driver, I can imagine wanting vendor std. Seems like a stretch though.

> Hypothetically (not sure I like it) topology attribs could be segregated from selectable attribs in the pool structure. Or they could be std.topo.pci or even topo.pcie ?

why? I think there are use cases to use `match` on things other than topo.

Reply:
> I don't follow. At the time we are considering whether a device fits this claim or not, we already have a device in mind, and so we have a total set of the attributes. If you care about which vendor, you should make your first criterion device.driver == "example.com". If you don't care about driver, then you clearly cannot match on vendor-defined attributes.

Suppose a device from vendor "dra.example.com" has these attributes:
- memory (= memory.dra.example.com, defined by vendor)
- gpu-type (= gpu-type.dra.example.com,  defined by vendor)
- pciRoot.dra.k8s.io (defined by Kubernetes in DRA KEP)
- gpu-type.cloud-native-ai.cncf.io (defined by the TAG, but a bit differently than done previously by that vendor).

How do we expose that third attribute if in CEL we only have `device.std` and `device.vendor`?

> There are only standard attributes and vendor attributes. 

That's the key point: I can easily imagine the desire from standard organizations or industry consortia to define their own set of attributes to encourage interoperability. We don't want those to be part of the std Kubernetes attributes (too many, name conflicts, too long lead time to role them out), nor should they fall under vendor (might conflict with names chosen by a vendor earlier).

Kubernetes itself might want to do the same. We can use `dra.k8s.io` for the DRA KEP and something else for other KEPs.

I stand by [my proposal](https://github.com/kubernetes-sigs/wg-device-management/pull/14#discussion_r1616631959). I'm okay with bringing back `DriverName` to keep CEL expressions shorter, but would prefer if we didn't because I prefer clarity (fully qualified names) over brevity (short names with implicit determination of the intended suffix).

Reply:
> Model isn't important, but I think driver and vendor make sense

Are we going to API-reject Pools with devices which lack these?

> I can easily imagine the desire from standard organizations

I can imagine all sorts of things, but I don't want to corrupt the design to accomodate my imagination.

That said, I understand what you're trying to accomodate.  I'm just a little skeptical.  IMO: common things should be simple.  I expect many more vendor-defined attributes than standard ones, which would suggest to me that vendor attributes must be "simple".

1) Are we OK with CEL saying `attribs["attr-name"]` (or a typed form of this) vs. `dev.attr_name` ?  This is material on the design.

Maybe we should get a consult from CEL experts, I feel like I am making potentially dubious assumptions.

Reply:
John and I had a quick chat with Joe and Cici on CEL.  First, what we're doing is not without precedent overall and "should work", so that's good.  That said, there's not really a "normal" way of doing this.

A few key points:
* There's not really a way to have complex field names (e.g. with dots), so any scoping must be reflected in structure.
* A map-type object supports both indexed lookups (`attrs["name"]`) and field lookups (`attrs.name`) at the same time, as long as the `name` is a valid identifier.
* What's left is mostly aesthetic.

Here are most (?) of the variations we have talked about:

1) `dev.vendorAttr` and `dev["standard-attr.k8s.io"]` - simple things are simple
2) `attrs["vendorAttr"]` and `attrs["standard-attr.k8s.io"]` - consistent
3) `stringAttrs["vendorAttr"]` and `boolAttrs["standard-attr.k8s.io"]` - typesafe(r) but verbose
4) `dev.vendorAttr` and `dev.x["k8s.io"].standardAttr` - feels more native
5) `attrs["driver"].vendorAttr` and `attrs["k8s.io"].standardAttr` - more consistent

Reply:
The problem with making "vendor" attributes "simple" (= `device.vendor.memory`) is that the code which evaluates the CEL expression doesn't know what "the vendor" is.

A CEL expression could look like this: `device.driver == "dra.example.com" && device.vendor.memory.isGreaterThan("1Gi") || device.driver == "other.example.com" && device.vendor.gpuMemory.isGreatherThan("1Gi")`. This is for the case where the app is fine with GPUs from different vendors.

In the first clause, `memory` = `memory.dra.example.com`. In the second clause, `gpuMemory` = `gpuMemory.other.example.com`. This is not just hard to figure out for the code, it can also be confusing for humans.

This is syntactic sugar. Can we start with the current proposal and add "nicer" CEL expressions as a future extension?

I will make one change: validation must ensure that the first part of the attribute name is a C-identifier (underscores allowed, hyphen not, no leading digits). Then we can use it safely as a field name in the future.

Reply:
> evaluates the CEL expression doesn't know what "the vendor" is.

We would have separate []Attribute slices in the Device, one for "vendor" and one for "standard".

Reply:
@johnbelamaric: it is not about "this attribute is a vendor attribute". It is about "this is a vendor attribute *of vendor example.com*". When the user writes `device.vendor.memory`, do they want to look up `memory.dra.example.com` or `memory.other.example.com`?

Those could be different things. Depending on the vendor of the current device, one or the other lookup should fail.

Reply:
To make this even more obvious:

```yaml
kind: resourceClaim
...
spec:
  requests:
  - deviceSelector: device.memory.greaterThan(quantity("1Gi"))
```

Should this really look up any vendor attribute called "memory", even if two vendors don't agree on what "memory" means?

Reply:
When we discussed this, the conclusion was that its the responsibility of the user to ensure that `device.memory` (or whatever the actual syntax will be) is matching against a device of the expected vendor. Often this will be done by referencing a DeviceClass.